contracts/CryptoMarriage.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.7;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721BurnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+
+contract CryptoMarriage is
+  Initializable,
+  ERC721Upgradeable,
+  ERC721BurnableUpgradeable,
+  AccessControlUpgradeable
+{
+  using CountersUpgradeable for CountersUpgradeable.Counter;
+
+  struct Marriage {
+    uint256 referenceTime;
+    uint256[] dayOffsets;
+    string[] uris;
+  }
+
+  bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+  mapping(uint256 => Marriage) public marriages;
+
+  CountersUpgradeable.Counter private _tokenIdCounter;
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() initializer {}
+
+  function initialize() public initializer {
+    __ERC721_init("CryptoMarriage", "CM");
+    __ERC721Burnable_init();
+    __AccessControl_init();
+
+    _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    _grantRole(MINTER_ROLE, msg.sender);
+  }
+
+  function createMarriage(
+    address to,
+    uint256 _referenceTime,
+    string[] calldata _uris,
+    uint256[] calldata _daysOffsets
+  ) public {
+    require(
+      hasRole(MINTER_ROLE, msg.sender),
+      "Only minters can create marriages"
+    );
+    require(
+      _referenceTime < block.timestamp,
+      "Reference time must be in the past"
+    );
+    uint256 tokenId = _tokenIdCounter.current();
+    _tokenIdCounter.increment();
+    _safeMint(to, tokenId);
+
+    require(
+      _uris.length == _daysOffsets.length + 1,
+      "Number of uris must be equal to number of day offsets + 1"
+    );
+
+    require(
+      _daysOffsets.length > 0,
+      "Number of day offsets must be greater than 0"
+    );
+
+    if (_daysOffsets.length > 1) {
+      for (uint256 index = 1; index < _daysOffsets.length; index++) {
+        require(
+          _daysOffsets[index] > _daysOffsets[index - 1],
```

Comment:
isn't the main use case for this to change every year? Could probably simplify a couple things a bit

Reply:
The one I want to mint won't evolve linearly. I was thinking every few months in the beginning, then yearly for a while then every 5 years around the end.
I think it could be useful to have the option to configure it for each nft

Reply:
"around the end", sounds ominous 😅

Reply:
It's actually the opposite, when you reach the end, [your marriage is made of Iridium, the densest metal on earth](https://ipfs.io/ipfs/QmNsNJ6GTtq6u2y2Vi5Zkf9yk2D24w9hojLxvv6C83LRwp)